File: tools/convert_hf.py
#!/usr/bin/env python3
"""Convert Qwen3-0.6B HuggingFace safetensors → ANE checkpoint format.

Usage:
    python3 convert_hf.py [--model Qwen/Qwen3-0.6B] [--output ane_qwen3_06b_dyn_ckpt.bin]

Downloads the model from HuggingFace (cached), converts weight layout,
writes the binary checkpoint that ./infer can load directly.
"""
import argparse
import struct
import numpy as np
import sys
import os

# Qwen3-0.6B architecture constants (must match models/qwen3_06b.h)
DIM = 1024
HIDDEN = 3072
HEADS = 16
KV_HEADS = 8
HD = 128
Q_DIM = HEADS * HD      # 2048
KV_DIM = KV_HEADS * HD  # 1024
SEQ = 256
NLAYERS = 28
VOCAB = 151936

# Derived weight sizes
WQ_SZ = Q_DIM * DIM
WK_SZ = KV_DIM * DIM
WV_SZ = KV_DIM * DIM
WO_SZ = DIM * Q_DIM
W1_SZ = HIDDEN * DIM
W2_SZ = DIM * HIDDEN
W3_SZ = HIDDEN * DIM


def main():
    parser = argparse.ArgumentParser(description="Convert Qwen3-0.6B HF → ANE checkpoint")
    parser.add_argument("--model", default="Qwen/Qwen3-0.6B", help="HuggingFace model ID")
    parser.add_argument("--output", default="ane_qwen3_06b_dyn_ckpt.bin", help="Output checkpoint path")
    args = parser.parse_args()

    try:
        from safetensors import safe_open
    except ImportError:
        print("pip install safetensors", file=sys.stderr)
        sys.exit(1)

    try:
        from huggingface_hub import snapshot_download
    except ImportError:
        print("pip install huggingface_hub", file=sys.stderr)
        sys.exit(1)

    # Download model
    print(f"Downloading {args.model}...")
    model_path = snapshot_download(args.model)
    print(f"Model cached at: {model_path}")

    # Find safetensors files
    st_files = sorted([f for f in os.listdir(model_path) if f.endswith(".safetensors")])
    if not st_files:
        print("No .safetensors files found!", file=sys.stderr)
        sys.exit(1)
    print(f"Found {len(st_files)} safetensors file(s)")

    # Load all tensors (Qwen3 uses bfloat16 — need torch for bf16→f32)
    try:
        import torch
        use_torch = True
    except ImportError:
        use_torch = False

    tensors = {}
    for sf in st_files:
        path = os.path.join(model_path, sf)
        if use_torch:
            with safe_open(path, framework="pt") as f:
                for key in f.keys():
                    tensors[key] = f.get_tensor(key).float().numpy()
        else:
            # Fallback: read raw bf16 bytes, manual convert via struct
            with safe_open(path, framework="numpy", device="cpu") as f:
                for key in f.keys():
                    tensors[key] = f.get_tensor(key)
    print(f"Loaded {len(tensors)} tensors")

    # Print tensor names for debugging
    for k, v in sorted(tensors.items()):
        print(f"  {k}: {v.shape} {v.dtype}")

    # Verify shapes
    # Qwen3 naming: model.layers.{i}.self_attn.{q,k,v,o}_proj.weight
    #               model.layers.{i}.mlp.{gate,up,down}_proj.weight
    #               model.layers.{i}.input_layernorm.weight
    #               model.layers.{i}.post_attention_layernorm.weight
    #               model.norm.weight
    #               model.embed_tokens.weight

    def get(name):
        t = tensors[name].astype(np.float32)
        return t

    # Write checkpoint
    print(f"\nWriting checkpoint: {args.output}")
    with open(args.output, "wb") as f:
        # CkptHdr (must match config.h CkptHdr struct exactly)
        # struct CkptHdr {
        #   int magic, version, step, total_steps;
        #   int n_layers, vocab_size, dim, hidden_dim, n_heads, seq_len;
        #   float lr, loss;
        #   double cum_compile, cum_train, cum_wall;
        #   int cum_steps, cum_batches, adam_t;
        #   int kv_heads, head_dim, q_dim;
        # }
        header = struct.pack(
            "iiii"      # magic, version, step, total_steps
            "iiiiii"    # n_layers, vocab_size, dim, hidden_dim, n_heads, seq_len
            "ff"        # lr, loss
            "ddd"       # cum_compile, cum_train, cum_wall
            "iii"       # cum_steps, cum_batches, adam_t
            "iii",      # kv_heads, head_dim, q_dim
            0x424C5A54, 5, 0, 0,          # magic, version=5 (QK-norm), step=0, total_steps=0
            NLAYERS, VOCAB, DIM, HIDDEN, HEADS, SEQ,
            0.0, 0.0,                      # lr, loss (not training)
            0.0, 0.0, 0.0,                # cum_compile, cum_train, cum_wall
            0, 0, 0,                       # cum_steps, cum_batches, adam_t
            KV_HEADS, HD, Q_DIM,
        )
        f.write(header)
        print(f"  Header: {len(header)} bytes")

        # Per-layer weights + zero Adam state
        adam_zeros_per_weight = None  # will compute per tensor
        for L in range(NLAYERS):
            prefix = f"model.layers.{L}"

            # Qwen3 attention: q_proj [Q_DIM, DIM], k_proj [KV_DIM, DIM], etc.
            Wq = get(f"{prefix}.self_attn.q_proj.weight")  # [Q_DIM, DIM]
            Wk = get(f"{prefix}.self_attn.k_proj.weight")  # [KV_DIM, DIM]
            Wv = get(f"{prefix}.self_attn.v_proj.weight")  # [KV_DIM, DIM]
            Wo = get(f"{prefix}.self_attn.o_proj.weight")  # [DIM, Q_DIM]

            # Qwen3 MLP: gate_proj=W1, down_proj=W2, up_proj=W3
            W1 = get(f"{prefix}.mlp.gate_proj.weight")     # [HIDDEN, DIM]
            W2 = get(f"{prefix}.mlp.down_proj.weight")     # [DIM, HIDDEN]
            W3 = get(f"{prefix}.mlp.up_proj.weight")       # [HIDDEN, DIM]

            rms_att = get(f"{prefix}.input_layernorm.weight")           # [DIM]
            rms_ffn = get(f"{prefix}.post_attention_layernorm.weight")  # [DIM]

            # QK-norm weights (Qwen3-specific)
            q_norm_w = get(f"{prefix}.self_attn.q_norm.weight")        # [HD]
            k_norm_w = get(f"{prefix}.self_attn.k_norm.weight")        # [HD]

            # Verify shapes
            assert Wq.shape == (Q_DIM, DIM), f"L{L} Wq: {Wq.shape} != ({Q_DIM}, {DIM})"
            assert Wk.shape == (KV_DIM, DIM), f"L{L} Wk: {Wk.shape} != ({KV_DIM}, {DIM})"
            assert Wv.shape == (KV_DIM, DIM), f"L{L} Wv: {Wv.shape} != ({KV_DIM}, {DIM})"
            assert Wo.shape == (DIM, Q_DIM), f"L{L} Wo: {Wo.shape} != ({DIM}, {Q_DIM})"
            assert W1.shape == (HIDDEN, DIM), f"L{L} W1: {W1.shape} != ({HIDDEN}, {DIM})"
            assert W2.shape == (DIM, HIDDEN), f"L{L} W2: {W2.shape} != ({DIM}, {HIDDEN})"
            assert W3.shape == (HIDDEN, DIM), f"L{L} W3: {W3.shape} != ({HIDDEN}, {DIM})"
            assert rms_att.shape == (DIM,), f"L{L} rms_att: {rms_att.shape}"
            assert rms_ffn.shape == (DIM,), f"L{L} rms_ffn: {rms_ffn.shape}"
            assert q_norm_w.shape == (HD,), f"L{L} q_norm: {q_norm_w.shape} != ({HD},)"
            assert k_norm_w.shape == (HD,), f"L{L} k_norm: {k_norm_w.shape} != ({HD},)"

            # Write weights (row-major float32, matching save_checkpoint order)
            f.write(Wq.tobytes())
            f.write(Wk.tobytes())
            f.write(Wv.tobytes())
            f.write(Wo.tobytes())
            f.write(W1.tobytes())
            f.write(W2.tobytes())
            f.write(W3.tobytes())
            f.write(rms_att.tobytes())
            f.write(rms_ffn.tobytes())
            f.write(q_norm_w.tobytes())
            f.write(k_norm_w.tobytes())

            # Adam state: 2 floats (m,v) per weight element + 2 rms + 2 qk-norm
            # Order: Wq(m,v), Wk(m,v), Wv(m,v), Wo(m,v), W1(m,v), W2(m,v), W3(m,v), rms_att(m,v), rms_ffn(m,v), q_norm(m,v), k_norm(m,v)
            adam_sizes = [WQ_SZ, WK_SZ, WV_SZ, WO_SZ, W1_SZ, W2_SZ, W3_SZ, DIM, DIM, HD, HD]
            for sz in adam_sizes:
                f.write(np.zeros(sz, dtype=np.float32).tobytes())  # m
                f.write(np.zeros(sz, dtype=np.float32).tobytes())  # v

            if L % 7 == 0:
                print(f"  Layer {L}/{NLAYERS}")

        # rms_final (model.norm.weight)
        rms_final = get("model.norm.weight")  # [DIM]
        assert rms_final.shape == (DIM,)
        f.write(rms_final.tobytes())

        # Adam state for rms_final (m, v)
        f.write(np.zeros(DIM, dtype=np.float32).tobytes())
        f.write(np.zeros(DIM, dtype=np.float32).tobytes())

        # Embedding (model.embed_tokens.weight) [VOCAB, DIM]
        embed = get("model.embed_tokens.weight")
        assert embed.shape == (VOCAB, DIM), f"embed: {embed.shape} != ({VOCAB}, {DIM})"
        f.write(embed.tobytes())

        # Adam state for embed (m, v)
        f.write(np.zeros(VOCAB * DIM, dtype=np.float32).tobytes())
        f.write(np.zeros(VOCAB * DIM, dtype=np.float32).tobytes())

    file_size = os.path.getsize(args.output)
    print(f"\nDone! {args.output}: {file_size / 1e9:.2f} GB")
    print(f"Run inference: python3 infer.py --ckpt {args.output} \"Once upon a time\"")


if __name__ == "__main__":
    main()

File: tools/convert_hf.rs
//! Convert HuggingFace safetensors → ANE checkpoint
//! Usage: cargo run --release --bin convert_hf -- [--model Qwen/Qwen3-0.6B] [--output ckpt.bin]
//!
//! Downloads safetensors from HuggingFace, converts bf16→f32, writes ANE checkpoint.

use ane::config;
use safetensors::SafeTensors;
use std::io::Write;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = std::env::args().collect();
    let mut model_id = "Qwen/Qwen3-0.6B".to_string();
    let mut output = "ane_qwen3_06b_dyn_ckpt.bin".to_string();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--model" => { i += 1; model_id = args[i].clone(); }
            "--output" => { i += 1; output = args[i].clone(); }
            _ => {}
        }
        i += 1;
    }

    let cfg = config::qwen3_06b();
    eprintln!("Converting {} → {}", model_id, output);
    eprintln!("Config: dim={} q_dim={} kv_dim={} hidden={} heads={} kv_heads={} hd={} layers={} vocab={}",
        cfg.dim, cfg.q_dim, cfg.kv_dim, cfg.hidden, cfg.heads, cfg.kv_heads, cfg.hd, cfg.n_layers, cfg.vocab);

    // Download safetensors
    eprintln!("Downloading model...");
    let cache_dir = dirs_or_default();
    let st_path = download_safetensors(&model_id, &cache_dir)?;
    eprintln!("Safetensors: {}", st_path);

    let st_bytes = std::fs::read(&st_path)?;
    let st = SafeTensors::deserialize(&st_bytes)?;
    eprintln!("Loaded {} tensors", st.names().len());

    let mut f = std::fs::File::create(&output)?;

    // Write CkptHdr (88 bytes)
    let header = CkptHeader {
        magic: 0x424C5A54, version: 5, step: 0, total_steps: 0,
        n_layers: cfg.n_layers as i32, vocab_size: cfg.vocab as i32,
        dim: cfg.dim as i32, hidden_dim: cfg.hidden as i32,
        n_heads: cfg.heads as i32, seq_len: cfg.seq as i32,
        lr: 0.0, loss: 0.0,
        cum_compile: 0.0, cum_train: 0.0, cum_wall: 0.0,
        cum_steps: 0, cum_batches: 0, adam_t: 0,
        kv_heads: cfg.kv_heads as i32, head_dim: cfg.hd as i32, q_dim: cfg.q_dim as i32,
    };
    let hdr_bytes = unsafe {
        std::slice::from_raw_parts(&header as *const CkptHeader as *const u8, std::mem::size_of::<CkptHeader>())
    };
    f.write_all(hdr_bytes)?;

    // Per-layer
    let adam_sizes = [cfg.wq_size(), cfg.wk_size(), cfg.wv_size(), cfg.wo_size(),
        cfg.w1_size(), cfg.w2_size(), cfg.w3_size(), cfg.dim, cfg.dim, cfg.hd, cfg.hd];

    for l in 0..cfg.n_layers {
        let pfx = format!("model.layers.{l}");
        write_tensor(&mut f, &st, &format!("{pfx}.self_attn.q_proj.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.self_attn.k_proj.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.self_attn.v_proj.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.self_attn.o_proj.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.mlp.gate_proj.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.mlp.down_proj.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.mlp.up_proj.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.input_layernorm.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.post_attention_layernorm.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.self_attn.q_norm.weight"))?;
        write_tensor(&mut f, &st, &format!("{pfx}.self_attn.k_norm.weight"))?;

        // Zero Adam state (m, v for each weight)
        for &sz in &adam_sizes {
            let zeros = vec![0u8; sz * 4 * 2]; // m + v
            f.write_all(&zeros)?;
        }
        if l % 7 == 0 { eprintln!("  Layer {l}/{}", cfg.n_layers); }
    }

    // rms_final + Adam
    write_tensor(&mut f, &st, "model.norm.weight")?;
    f.write_all(&vec![0u8; cfg.dim * 4 * 2])?;

    // Embedding + Adam
    write_tensor(&mut f, &st, "model.embed_tokens.weight")?;
    f.write_all(&vec![0u8; cfg.vocab * cfg.dim * 4 * 2])?;

    let size = std::fs::metadata(&output)?.len();
    eprintln!("\nDone! {}: {:.2} GB", output, size as f64 / 1e9);
    Ok(())
}

/// Convert tensor from safetensors (bf16/f16/f32) → f32 bytes, write to file
fn write_tensor(f: &mut std::fs::File, st: &SafeTensors, name: &str) -> Result<(), Box<dyn std::error::Error>> {
    let view = st.tensor(name).map_err(|e| format!("{}: {}", name, e))?;
    let data = view.data();
    let dtype = view.dtype();

    match dtype {
        safetensors::Dtype::BF16 => {
            let n = data.len() / 2;
            let mut f32_buf = vec![0.0f32; n];
            for i in 0..n {
                let bits = u16::from_le_bytes([data[2*i], data[2*i+1]]);
                f32_buf[i] = bf16_to_f32(bits);
            }
            let bytes = unsafe { std::slice::from_raw_parts(f32_buf.as_ptr() as *const u8, n * 4) };
            f.write_all(bytes)?;
        }
        safetensors::Dtype::F16 => {
            let n = data.len() / 2;
            let mut f32_buf = vec![0.0f32; n];
            for i in 0..n {
                let bits = u16::from_le_bytes([data[2*i], data[2*i+1]]);
                f32_buf[i] = ane::surface::fp16_to_f32(bits);
            }
            let bytes = unsafe { std::slice::from_raw_parts(f32_buf.as_ptr() as *const u8, n * 4) };
            f.write_all(bytes)?;
        }
        safetensors::Dtype::F32 => {
            f.write_all(data)?;
        }
        _ => return Err(format!("{}: unsupported dtype {:?}", name, dtype).into()),
    }
    Ok(())
}

fn bf16_to_f32(bits: u16) -> f32 {
    f32::from_bits((bits as u32) << 16)
}

#[repr(C)]
struct CkptHeader {
    magic: u32, version: u32, step: i32, total_steps: i32,
    n_layers: i32, vocab_size: i32, dim: i32, hidden_dim: i32,
    n_heads: i32, seq_len: i32, lr: f32, loss: f32,
    cum_compile: f64, cum_train: f64, cum_wall: f64,
    cum_steps: i32, cum_batches: i32, adam_t: i32,
    kv_heads: i32, head_dim: i32, q_dim: i32,
}

fn dirs_or_default() -> String {
    std::env::var("HF_HOME").unwrap_or_else(|_| {
        format!("{}/.cache/huggingface", std::env::var("HOME").unwrap_or_default())
    })
}

fn download_safetensors(model_id: &str, _cache_dir: &str) -> Result<String, Box<dyn std::error::Error>> {
    // Check if already cached
    let cache_model = format!("{}/.cache/huggingface/hub/models--{}",
        std::env::var("HOME").unwrap_or_default(),
        model_id.replace('/', "--"));

    // Look for safetensors in cache
    if let Ok(entries) = std::fs::read_dir(format!("{}/snapshots", cache_model)) {
        for entry in entries.flatten() {
            let snap_dir = entry.path();
            if let Ok(files) = std::fs::read_dir(&snap_dir) {
                for f in files.flatten() {
                    let name = f.file_name().to_string_lossy().to_string();
                    if name.ends_with(".safetensors") {
                        return Ok(f.path().to_string_lossy().to_string());
                    }
                }
            }
        }
    }

    // Download via HuggingFace API
    eprintln!("Downloading from HuggingFace API...");
    let url = format!("https://huggingface.co/{}/resolve/main/model.safetensors", model_id);
    let out_path = format!("{}.safetensors", model_id.replace('/', "_"));

    let resp = ureq::get(&url).call()?;
    let mut reader = resp.into_body().into_reader();
    let mut out = std::fs::File::create(&out_path)?;
    std::io::copy(&mut reader, &mut out)?;

    Ok(out_path)
}
